pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go

```diff
@@ -0,0 +1,209 @@
+package extendedsecret
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	certv1 "k8s.io/api/certificates/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	certv1client "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	esv1 "code.cloudfoundry.org/cf-operator/pkg/kube/apis/extendedsecret/v1alpha1"
+	"code.cloudfoundry.org/cf-operator/pkg/kube/util/config"
+	"code.cloudfoundry.org/cf-operator/pkg/kube/util/ctxlog"
+	"code.cloudfoundry.org/cf-operator/pkg/kube/util/names"
+)
+
+// NewCertificateSigningRequestReconciler returns a new Reconciler
+func NewCertificateSigningRequestReconciler(ctx context.Context, config *config.Config, mgr manager.Manager, certClient certv1client.CertificatesV1beta1Interface, srf setReferenceFunc) reconcile.Reconciler {
+	return &ReconcileCertificateSigningRequest{
+		ctx:          ctx,
+		config:       config,
+		client:       mgr.GetClient(),
+		certClient:   certClient,
+		scheme:       mgr.GetScheme(),
+		setReference: srf,
+	}
+}
+
+// ReconcileCertificateSigningRequest reconciles an CertificateSigningRequest object
+type ReconcileCertificateSigningRequest struct {
+	ctx          context.Context
+	config       *config.Config
+	client       client.Client
+	certClient   certv1client.CertificatesV1beta1Interface
+	scheme       *runtime.Scheme
+	setReference setReferenceFunc
+}
+
+// Reconcile approves pending CSR and creates its certificate secret
+func (r *ReconcileCertificateSigningRequest) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	instance := &certv1.CertificateSigningRequest{}
+
+	// Set the ctx to be Background, as the top-level context for incoming requests.
+	ctx, cancel := context.WithTimeout(r.ctx, r.config.CtxTimeOut)
+	defer cancel()
+
+	ctxlog.Infof(ctx, "Reconciling certificateSigningRequest '%s'", request.Name)
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			ctxlog.Info(ctx, "Skip reconcile: certificateSigningRequest not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		ctxlog.Info(ctx, "Error reading the object")
+		return reconcile.Result{}, errors.Wrap(err, "Error reading certificateSigningRequest")
+	}
+
+	if len(instance.Status.Certificate) != 0 {
+		annotations := instance.GetAnnotations()
```

Comment:
I'm just wondering, why do we need these annotations? Couldn't we get the secret name and namespace from the owner object (ExtendedSecret)?

Reply:
The truth is that:
1. CSR is no-namespaced
2. the owner and owned object are both namespace-scoped and in the same namespace

So getting the owner's namespace from CSR could not get right namespace. 
My idea is to set the owner's namespace to annotations.